### PR TITLE
Dynamic dispatch for Keras backend

### PR DESCRIPTION
**Pull request by bstriner**
Hi everybody!

This PR provides dynamic dispatch in the Keras backend. Python reflection magic.

TL;DR: Instant support for any function in the theano or tensorflow backend not already in Keras backend, like arccos, etc.

* If a function is defined in the keras backend, `K.method` works as before.
* If a function is not defined in the keras backend, but defined in the backend module, `K.method` will call `tf.method` or `theano.tensor.method` depending on the backend module.
* If a function is not in keras backend or the backend module, it will throw a custom message `AttributeError` that the requested function is not in either the keras backend or the selected backend module

Many functions have the same name and invocation in theano and tensorflow. When a new function comes out in either library, we can write code like `K.method`. If eventually someone writes some custom code in the keras backend, `K.method` will start calling that custom code.

Hypothetically, could delete large portions of the backend after this pull request, but might cause performance and documentation issues.

Shouldn't impact any existing code but will wait for tests to complete before celebrating.

Cheers,
Ben

### CODE REVIEW

**Comment by bstriner** [comment]
So I did some testing and dynamic dispatch works except for messing up `dir(keras.backend)`. Here is an alternative that just copies any functions that aren't already in the backend into the backend. This version works correctly with `dir()`.

```
#Copy any functions from the selected backend not already in keras.backend
for func in dir(backend()):
    if not hasattr(globals(), func):
        globals()[func]=getattr(backend(), func)
```

**Comment by bstriner** [comment]
For anyone interested in dynamic dispatch from an academic perspective, this code works other than `dir` doesn't list all functions.

```
class Backend:
  def __init__(self, module):
    self.module = module

  def __getattr__(self, item):
    if hasattr(self.module, item):
      return getattr(self.module, item)
    elif hasattr(self.module.get_backend(), item):
      return getattr(self.module.get_backend(), item)
    else:
      raise AttributeError(
             "Neither module 'keras.backend' nor '{}' has attribute '{}'".format(self.module.backend(), item))

sys.modules[__name__] = Backend(sys.modules[__name__])
```

**Comment by bstriner** [comment]
These are the functions added in Theano. Most won't be useful but some like jacobian or arcsinh might be useful.

```
ARange
AdvancedIncSubtensor
AdvancedIncSubtensor1
AdvancedIndexingError
AdvancedSubtensor
AdvancedSubtensor1
Alloc
AllocEmpty
Apply
Argmax
AsTensorError
CAReduce
Choose
Constant
Default
Diag
DiffOp
DimShuffle
DisconnectedType
Dot
Elemwise
EmptyConstantError
ExtractDiag
Eye
Flatten
Generic
IncSubtensor
Join
LoadFromDisk
Lop
MPIRecv
MPIRecvWait
MPISend
MPISendWait
MakeSlice
MaxAndArgmax
Mean
MethodNotDefined
NoneConst
NoneTypeT
Nonzero
NotScalarConstantError
Op
PermuteRowElements
Rebroadcast
Reshape
Rop
ScalarFromTensor
Shape
ShapeError
SliceConstant
SliceType
SpecifyShape
Split
Subtensor
SubtensorPrinter
Sum
Tensor
TensorConstant
TensorConstantSignature
TensorFromScalar
TensorType
TensorVariable
Tile
Tri
Type
Variable
__docformat__
_shared
_tensor_py_operators
abs_
add
addbroadcast
adv_index_broadcastable_pattern
advanced_inc_subtensor
advanced_inc_subtensor1
advanced_set_subtensor
advanced_set_subtensor1
advanced_subtensor
advanced_subtensor1
all_dtypes
allclose
alloc
alloc_validate_shape
and_
angle
arccos
arccosh
arcsin
arcsinh
arctan
arctan2
arctanh
argsort
as_index_variable
as_int_none_variable
as_tensor
as_tensor_variable
bartlett
basic
batched_dot
batched_tensordot
bcol
bincount
bitwise_and
bitwise_not
bitwise_or
bitwise_xor
blas
blas_c
blas_headers
blas_scipy
bmatrix
brow
bscalar
btensor3
btensor4
btensor5
builtins
bvector
ccol
ceil
ceil_intdiv
check_equal_numpy
chi2sf
choose
cmatrix
col
cols
compile
complex
complex_dtypes
complex_from_polar
complex_matrix_types
complex_scalar_types
complex_types
complex_vector_types
config
conj
consider_constant
constant
constant_cache
constant_or_value
constructor
continuous_dtypes
cosh
crow
cscalar
ctensor3
ctensor4
ctensor5
cumprod
cumsum
cvector
dcol
dcols
dedent
default
deg2rad
diag
diagonal
discrete_dtypes
div_proxy
division
divmod
dmatrices
dmatrix
drow
drows
dscalar
dscalars
dtensor3
dtensor3s
dtensor4
dtensor4s
dtensor5
dtensor5s
dvector
dvectors
elemwise
elemwise_cgen
eq
erf
erfc
erfcinv
erfcx
erfinv
exp2
expm1
extra_ops
extract_constant
fcol
fcols
fft
fill
fill_diagonal
fill_diagonal_offset
flatnonzero
float16_atol
float16_rtol
float32_atol
float32_rtol
float64_atol
float64_rtol
float_dtypes
float_matrix_types
float_scalar_types
float_types
float_vector_types
floor
floor_div
fmatrices
fmatrix
frow
frows
fscalar
fscalars
ftensor3
ftensor3s
ftensor4
ftensor4s
ftensor5
ftensor5s
fvector
fvectors
gamma
gammaln
ge
get_canonical_form_slice
get_idx_list
get_scalar_constant_value
get_scalar_constant_value_elemwises
get_vector_length
gof
grad
grad_not_implemented
grad_undefined
gt
hashtype
hessian
horizontal_stack
icol
icols
identity_like
imag
imatrices
imatrix
inc_subtensor
inplace_increment
int_div
int_dtypes
int_matrix_types
int_scalar_types
int_types
int_vector_types
integer_dtypes
integer_types
inv
inverse_permutation
invert
io
irecv
iround
irow
irows
is_flat
iscalar
iscalars
isclose
isend
isinf
isinf_
isnan
isnan_
itensor3
itensor3s
itensor4
itensor4s
itensor5
itensor5s
ivector
ivectors
izip
j0
j1
jacobian
join
key_to_cmp
largest
lcol
lcols
le
lmatrices
lmatrix
load
log10
log1p
log2
logging
lrow
lrows
lscalar
lscalars
lt
ltensor3
ltensor3s
ltensor4
ltensor4s
ltensor5
ltensor5s
lvector
lvectors
makeKeepDims
make_constant
make_slice
matrices
matrix
max_and_argmax
mgrid
min_informative_str
mod
mod_check
mpi_cmps
mpi_enabled
mpi_keys
mpi_send_wait_cmp
mpi_send_wait_key
mpi_tag_cmp
mpi_tag_key
mul
neg
neq
nlinalg
nnet
none_type_t
nonzero
nonzero_values
numbers
numeric_grad
numpy
numpy_diagonal_return_view
numpy_scalar
ogrid
opt
opt_uncanonicalize
or_
outer
partial
patternbroadcast
permute_row_elements
power
pprint
printing
psi
ptp
python_all
python_any
python_complex
rad2deg
raw_random
real
recv
register_transfer
roll
round_half_away_from_zero
round_half_to_even
row
rows
scal
scalar
scalar_from_tensor
scalars
second
send
set_subtensor
setdefault
sgn
shape_padaxis
shape_padleft
shape_padright
shared_randomstreams
sharedvar
signal
sinh
slicetype
slinalg
smallest
sort
sparse_module_ref
specify_shape
split
sqr
stacklists
sub
subtensor
swapaxes
take
tan
tensor
tensor3
tensor3s
tensor4
tensor4s
tensor5
tensor5s
tensor_copy
tensor_from_scalar
tensordot
transfer
tri
tril
triu
true_div
trunc
type
type_other
uint_dtypes
unbroadcast
utils
values_eq_approx_always_true
vector
vectors
verify_grad
vertical_stack
warnings
wcol
where
wmatrix
wrow
wscalar
wtensor3
wtensor4
wtensor5
wvector
xlogx
xor
xrange
zcol
zmatrix
zrow
zscalar
ztensor3
ztensor4
ztensor5
zvector
```

**Comment by gvtulder** [comment]
> Interested in everyone's thoughts.

My first thought: clever. But isn't this a bit *too* clever?

As you write, most of the things in the Theano list won't be useful, but they are still polluting the namespace. I can also see how this could be confusing: right now, you know that (almost) every function you find in `keras.backend` works with both backends and will work in the same way. If you blindly copy all functions from `theano.tensor`, you can no longer be sure that it will also be there in TensorFlow, and you also can't be sure that it has the same meaning.

I do like the `get_backend()` function, though. Wouldn't that already solve most of the problem? Then you can still call `K.get_backend().cos` and it will work for both backends, but at least you know that you're calling backend-specific methods and should be careful.

**Comment by bstriner** [comment]
Thanks! All good points. TL;DR, what do you think about the code below as an alternative? It doesn't provide instant pass-through, but at least it makes it really easy to add new functions.

 What I should have said is more along these lines: I have no personal reason to use most of the functions on that list, but then again, most functions I could imagine someone somewhere finding useful at some point.

This is my ideal use case:

1. Tensorflow releases function `qwerty` at midnight and you can use `K.qwerty` at 12:01 without modifying keras.
2. Months down the road, theano releases `qwerty` and your original code can work on theano without any modification to keras or your code.
3. Keras finally adds a custom implementation of `qwerty` to the backend and it is faster and fixes some numerical stability issues in the tensorflow code, and your code will start using the keras customization without modification.

So, if you just have `get_backend`, `K.get_backend().arccos` will go directly to the backend. If the other backend didn't have the function, the call would fail. If you do the aliasing, you can be much more flexible. There may also be situations where a function is an alias in one backend but a custom function in the other.

Maybe a pattern like filtering leading and trailing underscores and capitals would take out some redundancy and clean up the list somewhat.

An interesting side effect is that `help(K.arccos)` will display the theano or tensorflow documentation, which is way more detailed than anything we could try to write.

Did some thinking and testing since yesterday and here is an example of another approach. Only copies in what we want, but does it efficiently. This might be what I'm leaning towards, because it isn't too radical but gets the job done:

```
#theano_backend.py
arccos, arcsin, arctan = theano.tensor.arccos,  theano.tensor.arcsin,  theano.tensor.arctan
```

Another alternative is to do something similar to the current state of the PR, but iterate over a list of stuff to copy.

Some discussion at:
https://github.com/fchollet/keras/pull/4939
https://github.com/fchollet/keras/issues/4929

Cheers,
Ben

**Comment by gvtulder** [comment]
Good point, I hadn't realised that `K.get_backend().something` doesn't give you the option to override functions.

I very much prefer your second option, explicitly aliasing functions, over just copying everything and hoping for the best. (But this is not for me to decide, of course.) If TensorFlow adds that hypothetical new function at midnight, you can still call `tf.new_function` directly. I must admit that I don't switch backends every day.

**Comment by bstriner** [comment]
@gvtulder forgot to mention the original version, which didn't pollute the namespace. Maybe I should go back to that one? Too many choices!

With a little reflection magic, if you call a function not in keras backend, it will check if that function is in tf/th before throwing the missing attribute exception. dir lists just the functions in keras.backend, but you can call any function in keras.backend or tf/th.

The downside of that approach was that arccos, etc. wouldn't be listed in dir(K), it would just magically work when you call it. But maybe that way is better after all?

Cheers,
Ben

**Comment by bstriner** [comment]
Of course you can always use theano or tensorflow directly in keras, but anything we can do to encourage platform-independent code would be awesome.

I code in windows on theano and do my jobs on an ubuntu server running theano or tensorflow depending on the occasion. Seems like everyone I work with these days is on tensorflow, so I've been trying to make things more compatible, but I prefer theano myself.

**Comment by bstriner** [comment]
For anyone coming in at this point, options so far:

- Copy everything in. Functions from tf or th are instantly available and listed using dir(keras.backend) but pollutes namespace.
- Dynamically call everything: Functions from tf or th are instantly available but not listed using dir(keras.backend).
- Explicitly alias: functions are available as they are aliased, but at least minimal amount of code to accomplish the task.

For all three of these options, help(keras.backend.qwerty) will show the detailed tf or th documentation and we don't have to waste time documenting in keras.

Anyone out there feel strongly one way or another?

**Comment by gvtulder** [comment]
@bstriner: Well, for me the main benefit of the `keras.backend` abstraction is that it provides a set of functions that are guaranteed to work across backends. Having a bunch of automatically copied functions that may or may not work consistently removes most of the value, then I might just as well call the Theano function directly. The value, in my opinion, is that someone checked that the two backends provide the same functionality. But I guess that might be different if you regularly switch backends.

**Comment by fchollet** [comment]
This is dangerous for 2 reasons:

- Some thing will work using `K` with one backend but not the other(s) because some methods are present in one of the underlying frameworks but not the others. Normally, using `K` should be a guarantee of portability (with as few caveats as possible). That's why it is better to revert to using `tf` / `theano.tensor` when functionality is not available in `K`.
- There is no guarantee that a method present both in `tf` and `theano.tensor` under the same name will have the same API. There *are* significant API discrepancies here and there. This is a guaranteed source of serious, hard-to-debug issues.

**Comment by bstriner** [comment]
@fchollet I'll start by saying thank you for an amazing library and we are all in your debt.

Point taken regarding compatibility. Is option 3 still reasonable? For something like `sin`, `exp`, or `arctan` where the keras implementation is or will-be just a pass-through, aliasing gets the job done with minimal code. `help` will display the detailed docs from the underlying system. For the simple functions, of which there are probably a good number, this should be an efficient implementation.

```
#tensorflow_backend.py
sin, cos, tan = tf.sin, tf.cos, tf.tan
asin, acos, atan = tf.asin, tf.acos, tf.atan
```
```
#theano_backend.py
sin, cos, tan = theano.tensor.sin, theano.tensor.cos, theano.tensor.tan
asin, acos, atan = theano.tensor.arcsin, theano.tensor.arccos, theano.tensor.arctan
```

etc. for any other functions that are just simple wrappers.

If not, we'll just proceed with the usual function definitions for the trig functions, but I was interested in thoughts regarding more elegant ways to handle pass-through functions.

Cheers,
Ben

**Comment by brortao** [comment]
I like that approach! Do the aliased functions still show up in the generated documentation?

(I'm not familiar with MkDocs, and I couldn't quite get it to work, so I'm asking here).

**Comment by bstriner** [comment]
Given https://github.com/fchollet/keras/issues/4944, maybe the cleanest approach is to get the contrib repo up and running, and maintain some modules like `keras.contrib.backend.trig` or something manually.

To standardize the documentation for both backends, I added the following to `__init__.py`. Should make everything look the same as it did before, except for the added functions. Also, this means the documentation will be the same regardless of the backend when you generate, whereas it used to be different.

```
sin.func_name, sin.func_doc = "sin", "Computes sin of x element-wise.\n"
cos.func_name, cos.func_doc = "cos", "Computes cos of x element-wise.\n"
tan.func_name, tan.func_doc = "tan", "Computes tan of x element-wise.\n"
asin.func_name, asin.func_doc = "asin", "Computes asin of x element-wise.\n"
acos.func_name, acos.func_doc = "acos", "Computes acos of x element-wise.\n"
atan.func_name, atan.func_doc = "atan", "Computes atan of x element-wise.\n"
```

At this point, I'll just consider this an interesting experiment and close the PR. The intention was to make it easier to add new functions, but once you are customizing all of the documentation it is almost as much code as just writing the functions as usual.

Cheers,
Ben
